beacon-chain/utils/shuffle.go

```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
+		pivot := bytesutil.FromBytes8(ph[:8]) % listSize
+
+		// Split up the for-loop in two:
+		//  1. Handle the part from 0 (incl) to pivot (incl). This is mirrored around (pivot / 2)
+		//  2. Handle the part from pivot (excl) to N (excl). This is mirrored around ((pivot / 2) + (size/2))
+		// The pivot defines a split in the array, with each of the splits mirroring their data within the split.
+		// Print out some example even/odd sized index lists, with some even/odd pivots,
+		//  and you can deduce how the mirroring works exactly.
+		// Note that the mirror is strict enough to not consider swapping the index @mirror with itself.
+		mirror := (pivot + 1) >> 1
+		// Since we are iterating through the "positions" in order, we can just repeat the hash every 256th position.
+		// No need to pre-compute every possible hash for efficiency like in the example code.
+		// We only need it consecutively (we are going through each in reverse order however, but same thing)
+		//
+		// spec: source = hash(seed + int_to_bytes1(round) + int_to_bytes4(position // 256))
+		// - seed is still in 0:32 (excl., 32 bytes)
+		// - round number is still in 32
+		// - mix in the position for randomness, except the last byte of it,
+		//     which will be used later to select a bit from the resulting hash.
+		// We start from the pivot position, and work back to the mirror position (of the part left to the pivot).
+		// This makes us process each pear exactly once (instead of unnecessarily twice, like in the spec)
+		binary.LittleEndian.PutUint32(buf[pivotViewSize:], uint32(pivot>>8))
+		source := hashutil.HashSha256(buf)
+		byteV := source[(pivot&0xff)>>3]
+		for i, j := uint64(0), pivot; i < mirror; i, j = i+1, j-1 {
+			// The pair is i,j. With j being the bigger of the two, hence the "position" identifier of the pair.
+			// Every 256th bit (aligned to j).
+			if j&0xff == 0xff {
+				// just overwrite the last part of the buffer, reuse the start (seed, round)
+				binary.LittleEndian.PutUint32(buf[pivotViewSize:], uint32(j>>8))
+				source = hashutil.HashSha256(buf)
+			}
+			// Same trick with byte retrieval. Only every 8th.
+			if j&0x7 == 0x7 {
+				byteV = source[(j&0xff)>>3]
+			}
+			bitV := (byteV >> (j & 0x7)) & 0x1
+
+			if bitV == 1 {
+				// swap the pair items
+				input[i], input[j] = input[j], input[i]
+			}
+		}
+		// Now repeat, but for the part after the pivot.
+		mirror = (pivot + listSize + 1) >> 1
+		end := listSize - 1
+		// Again, seed and round input is in place, just update the position.
+		// We start at the end, and work back to the mirror point.
+		// This makes us process each pear exactly once (instead of unnecessarily twice, like in the spec)
+		binary.LittleEndian.PutUint32(buf[pivotViewSize:], uint32(end>>8))
+		source = hashutil.HashSha256(buf)
+		byteV = source[(end&0xff)>>3]
+		for i, j := pivot+1, end; i < mirror; i, j = i+1, j-1 {
+			// Exact same thing (copy of above loop body)
```

Comment:
Copy? This is screaming for abstraction! 
DRY = Don't repeat yourself

Please refactor this code so it is abstract enough to have reusable code.

Reply:
thanks, fixed it